pkg/imgpkg/imageset/image_set.go

```diff
@@ -77,57 +79,135 @@ func (o *ImageSet) Import(imgOrIndexes []imagedesc.ImageOrIndex,
 	o.logger.WriteStr("importing %d images...\n", len(imgOrIndexes))
 	defer func() { o.logger.WriteStr("imported %d images\n", len(importedImages.All())) }()
 
-	errCh := make(chan error, len(imgOrIndexes))
 	importThrottle := util.NewThrottle(o.concurrency)
 
+	imageOrIndexesToWrite := map[regname.Reference]regremote.Taggable{}
+	var imageOrIndexesToWriteLock = &sync.Mutex{}
+	errCh := make(chan error, len(imgOrIndexes))
 	for _, item := range imgOrIndexes {
 		item := item // copy
 
 		go func() {
 			importThrottle.Take()
 			defer importThrottle.Done()
+			addImageOrImageIndexToTaggableMapForMultiWrite(item, errCh, importRepo, registry, imageOrIndexesToWrite, imageOrIndexesToWriteLock)
+		}()
+	}
 
-			existingRef, err := regname.NewDigest(item.Ref())
-			if err != nil {
-				errCh <- err
-				return
-			}
+	err := checkForAnyAsyncErrors(imgOrIndexes, errCh)
+	if err != nil {
+		return nil, err
+	}
 
-			importDigestRef, err := o.importImage(item, existingRef, importRepo, registry)
-			if err != nil {
-				errCh <- fmt.Errorf("Importing image %s: %s", existingRef.Name(), err)
-				return
-			}
+	err = registry.MultiWrite(imageOrIndexesToWrite, o.concurrency)
+	if err != nil {
+		return nil, err
+	}
 
-			var regImage regv1.Image
-			if item.Image != nil {
-				regImage = *item.Image
-			}
-			var regImageIndex regv1.ImageIndex
-			if item.Index != nil {
-				regImageIndex = *item.Index
-			}
-			importedImages.Add(ProcessedImage{
-				UnprocessedImageRef: UnprocessedImageRef{existingRef.Name(), item.Tag()},
-				DigestRef:           importDigestRef.Name(),
-				Image:               regImage,
-				ImageIndex:          regImageIndex,
-			})
-			errCh <- nil
+	errChVerifyImages := make(chan error, len(imgOrIndexes))
+	for _, item := range imgOrIndexes {
+		item := item // copy
+
+		go func() {
```

Comment:
instead of propagating errChVerifyImages to function, let's have verifyProcessedImages return an error and use channel as close to go func() instantiation as possible. this way we are not leaking async nature of what we are doing into verifyProcessedImages and its children.

Reply:
same applies to verifyProcessedImages() taking errCh

Reply:
I'll go ahead and make this change. Although I don't fully understand the concern here.